src/Lazy/concurrentPool.ts

```diff
@@ -0,0 +1,179 @@
+import { isAsyncIterable } from "../_internal/utils";
+import type { Reject, Resolve } from "../types/Utils";
+
+type Item<T> = {
+  fail?: unknown;
+  success:
+    | { value: undefined; done: true } // done
+    | { value: T; done: false };
+};
+
+/**
+ * @experimental concurrentPool
+ */
+function concurrentPool<A>(
+  length: number,
+  iterable: AsyncIterable<A>,
+): AsyncIterableIterator<A>;
+
+function concurrentPool<A>(
+  length: number,
+  iterable?: AsyncIterable<A>,
+): (iterable: AsyncIterable<A>) => AsyncIterableIterator<A>;
+
+function concurrentPool<A>(
+  length: number,
+  iterable?: AsyncIterable<A>,
+):
+  | AsyncIterableIterator<A>
+  | ((iterable: AsyncIterable<A>) => AsyncIterableIterator<A>) {
+  if (iterable === undefined) {
+    return (iterable) => {
+      return concurrentPool(length, iterable);
+    };
+  }
+
+  if (!Number.isFinite(length) || length <= 0) {
+    throw new RangeError("'length' must be positive integer");
+  }
+  if (!isAsyncIterable(iterable)) {
+    throw new TypeError("'iterable' must be type of AsyncIterable");
+  }
+
+  /**
+   * @TODO ppeeou: Considering id circulation
+   */
+  let fillId = 1;
+  let consumeId = 1;
+  let idleWorkers = length;
+
+  const consumer: [Resolve<A>, Reject, number][] = [];
+  const itemMap: Map<number, Item<A>> = new Map();
+  const iterator = iterable[Symbol.asyncIterator]();
+
+  let finished = false;
+  let prev = Promise.resolve();
+
+  async function fetchNextItem() {
+    if (idleWorkers <= 0) {
+      return;
+    }
+
+    idleWorkers--;
+    const id = fillId++;
+    try {
+      const item = {
+        success: await iterator.next(),
+      } as Item<A>;
+      idleWorkers++;
+
+      itemMap.set(id, item);
+      await chainItem(item);
+    } catch (error) {
+      idleWorkers++;
+      finished = true;
+
+      const item = {
+        success: { done: false },
+        fail: error,
+      } as Item<A>;
+
+      itemMap.set(id, item);
+      await chainItem(item);
+    }
+  }
+
+  async function chainItem(item: Item<A>) {
+    prev = prev.then(() => item).then(pullItem);
+    return item;
+  }
```

Comment:
Is there any reason to use the `async` keyword if await is not used inside chainItem?